---
title: "Retrieving fastq files, Quality Assessment and counting"
author: "Mark Dunning"
date: "25 July 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Command-line analysis


## Retrieve the fastq file from Sequencing Read Archive (SRA)

We can download a fastq file from the Short Read Archive, provided we know it's location, using a `wget` unix command

```{bash eval=FALSE}

### DO NOT RUN
wget ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByStudy/sra/SRP/SRP045/SRP045534/SRR1552444/SRR1552444.sra -P sra_data/
```


The ftp site for Sequencing read archive can be accessed at `ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByStudy/sra/SRP/`

from there you can navigate to the folder containing a particular sequencing run


## Extract the fastq

The `sra-toolkit` provides various utilities for dealing with files in this format, including converting to the more popular `fastq` format.

```{bash eval=FALSE}
## DO NOT RUN - it will take too long
fastq-dump sra_data/SRR1552444.sra -O fastq_files
```

```{bash}
gzip fastq_files/SRR1552444.fastq
```


## Run the fastqc tool

```{bash eval=FALSE}
fastqc fastq_files/SRR1552444.fastq.gz -o qc
```


## Collecting the QC with multiqc


```{bash}
multiqc qc
```


## Download reference transcripts

```{bash eval=FALSE}
wget ftp://ftp.ensembl.org/pub/release-91/fasta/mus_musculus/cdna/Mus_musculus.GRCm38.cdna.all.fa.gz -P ensembl_ref/
wget ftp://ftp.ensembl.org/pub/release-91/gtf/mus_musculus/Mus_musculus.GRCm38.91.chr.gtf.gz -P ensembl_ref

```

## Creating a salmon index

```{bash eval=FALSE}
salmon index -i salmon_ref/GRCm38 -t ensembl_ref/Mus_musculus.GRCm38.cdna.all.fa.gz
```

## Salmon quantification

```{bash eval=FALSE}
salmon quant -i salmon_ref/GRCm38 --libType A  \
--gcBias  -r fastq_files/SRR1552444.fastq.gz -o SRR1552444

```

```{bash eval=FALSE}
ls SRR1552444
head SRR1552444/quant.sf

```


## Running for all samples

```{bash}
mkdir -p salmon_quant/
for i in {44..55}
do
salmon quant -i salmon_ref/GRCm38 --libType A  \
--gcBias -r fastq_files/SRR15524${i}.fastq.gz -o salmon_quant/SRR15524${i}
done

```


# Alignment to the genome

```{bash}
mkdir aligned_reads
wget ftp://ftp.ensembl.org/pub/release-91/fasta/mus_musculus/dna/Mus_musculus.GRCm38.dna.chromosome.19.fa.gz
gunzip Mus_musculus.GRCm38.dna.chromosome.19.fa.gz
hisat2-build Mus_musculus.GRCm38.dna.chromosome.19.fa.gz GRCm38_chr19_hisat
hisat2 -q -x GRCm38_chr19 fastq_files/SRR1552444.fastq.gz -S aligned_reads/SRR1552444.sam


```


## References

[https://bioconductor.github.io/BiocWorkshops/rna-seq-data-analysis-with-deseq2.html](DESeq2 tutorial from Bioconductor 2018 conference)

