---
title: "RNA-seq Analysis in R"
subtitle: "Annotation and Visualisation of RNA-seq results"
author: "Mark Dunning"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output:
  html_notebook:
    toc: yes
    toc_float: yes
  html_document:
    toc: yes
    toc_float: yes
minutes: 300
layout: page
bibliography: ref.bib
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

**Original Authors: Belinda Phipson, Anna Trigos, Matt Ritchie, Maria Doyle, Harriet Dashnow, Charity Law**, **Stephane Ballereau, Oscar Rueda, Ashley Sawle**
Based on the course [RNAseq analysis in R](http://combine-australia.github.io/2016-05-11-RNAseq/) delivered on May 11/12th 2016

## Resources and data files

This material has been created using the following resources:  
http://www.statsci.org/smyth/pubs/QLedgeRPreprint.pdf [@Lun2016]  
http://monashbioinformaticsplatform.github.io/RNAseq-DE-analysis-with-R/99-RNAseq_DE_analysis_with_R.html  
http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html
https://bioconductor.github.io/BiocWorkshops/rna-seq-data-analysis-with-deseq2.html



Before starting this section, we will make sure we have all the relevant objects from the Differential Expression analysis present.

```{r}
suppressPackageStartupMessages(library(DESeq2))

load("Robjects/DE.Rdata")
load("Robjects/preprocessing.Rdata")
```

# Overview

- Visualising DE results
- Getting annotation
- Retrieving gene models
- Exporting browser traecks
- Visualising results with respect to genomic location



We can now have a list of genes ordered according to their evidence for being differentially-expressed.

```{r}
library(dplyr)
library(tibble)

results.status <- as.data.frame(results(de.mf,contrast=c("Status","lactation","virgin"))) %>% 
  rownames_to_column("ENSEMBL")
  

results.ordered <- arrange(results.status, padj)
head(results.ordered)
```

In `DESeq2`, the function plotMA shows the log2 fold changes attributable to a given variable over the mean of normalized counts for all the samples in the DESeqDataSet. Points will be colored red if the adjusted p value is less than 0.1. Points which fall out of the window are plotted as open triangles pointing either up or down.

The log2 fold change for a particular comparison is plotted on the y-axis and the average of the counts normalized by size factor is shown on the x-axis ("M" for minus, because a log ratio is equal to log minus log, and "A" for average). Each gene is represented with a dot. Genes with an adjusted p value below a threshold (here 0.1, the default) are shown in red.


```{r}
plotMA(results(de.mf,contrast=c("Status","lactation","virgin")))
```

***Note*** You may see an error message when trying to make the above MA plot. This could be because both `limma` and `DESeq2` have a function called `plotMA`, and R can sometimes pick the wrong function. To explictly use the `DESeq2` function you can use:-

```{r}
DESeq2:::plotMA(results(de.mf,contrast=c("Status","lactation","virgin")))
```

MA-plots often display a fanning-effect at the left-hand side (genes with low numbers of counts) due to the high variability of the measurements for these genes. For more informative visualization and more accurate ranking of genes by effect size (the log fold change may sometimes be referred to as an effect size), the `DESeq2` authors recommend "shrinking" the log fold-changes which is available in DESeq2â€™s `lfcShrink` function. This results in more stable fold change values. The p-values are unaffected.

```{r}
res_LvsV <- lfcShrink(de.mf,contrast=c("Status","lactation","virgin"))
plotMA(res_LvsV)
```

We will re-define our results object to use these new fold-changes.

```{r}
results.ordered <- as.data.frame(res_LvsV) %>% 
  rownames_to_column("ENSEMBL") %>% 
  arrange(padj)
head(results.ordered)
```

Another common plot for displaying the results of a differential expression analysis is a *volcano plot*

```{r}

plot(results.ordered$log2FoldChange, -log10(results.ordered$padj),
     pch=16,       
     xlab="Log Fold Change",
       ylab="Significance")
degenes <- which(results.ordered$padj < 0.05)
points(results.ordered$log2FoldChange[degenes],-log10(results.ordered$padj)[degenes],
       col="red",
       pch=16)
```


It can also be useful to examine the counts of reads for a single gene across the groups. A simple function for making this plot is `plotCounts`, which normalizes counts by sequencing depth and adds a pseudocount of 1/2 to allow for log scale plotting. The counts are grouped by the variables in  `intgroup`, where more than one variable can be specified. Here we specify the gene which had the smallest p value from the results table created above. You can select the gene to plot by rowname or by numeric index:-

```{r}
plotCounts(dds, "ENSMUSG00000000381",intgroup = c("Status"))
```


However, it is hard to assess the biological significance of such a gene without more information about . To perform such a task we need to map between the identifiers we have in the `DESeq2` output and more familiar names.


## Adding annotation to the DESeq2 results

There are a number of ways to add annotation, but we will demonstrate how to do this using the *org.Mm.eg.db* package. This package is one of several *organism-level* packages which are re-built every 6 months. These packages are listed on the [annotation section](http://bioconductor.org/packages/release/BiocViews.html#___AnnotationData) of the Bioconductor, and are installed in the same way as regular Bioconductor packages. An alternative approach is to use `biomaRt`, an interface to the [BioMart](http://www.biomart.org/) resource. BioMart is much more comprehensive, but the organism packages fit better into the Bioconductor workflow.


```{r eval=FALSE}
### Only execute when you need to install the package
source("http://www.bioconductor.org/biocLite.R")
biocLite("org.Mm.eg.db")
# For Human
biocLite("org.Hs.eg.db")
```

The packages are larger in size that Bioconductor software pacakges, but essentially they are databases that can be used to make *offline* queries. 

```{r message=FALSE}
library(org.Mm.eg.db)
```


First we need to decide what information we want. In order to see what we can extract we can run the `columns` function on the annotation database.

```{r}
columns(org.Mm.eg.db)
```

We are going to filter the database by a key or set of keys in order to extract the information we want. Valid names for the key can be retrieved with the `keytypes` function.

```{r}
keytypes(org.Mm.eg.db)
```

We should see `ENSEMBL`, which is the type of key we are going to use in this case. If we are unsure what values are acceptable for the key, we can check what keys are valid with `keys`

```{r}
keys(org.Mm.eg.db, keytype="ENSEMBL")[1:10]
```



For the top gene in our analysis the call to the function would be:-

```{r eval=FALSE}
select(org.Mm.eg.db, keys="ENSMUSG00000000381",
       keytype = "ENSEMBL",columns=c("SYMBOL","GENENAME")
)


### In case of errors, try
### AnnotationDBI::select(org.Mm.eg.db, keys="ENSMUSG00000000381",keytype = "ENSEMBL",columns=c("SYMBOL","GENENAME")



```



To annotate our results, we definitely want gene symbols and perhaps the full gene name. Let's build up our annotation information into a new data frame using the `select` function.

```{r}
anno <- select(org.Mm.eg.db,keys=results.ordered$ENSEMBL,
              columns=c("SYMBOL","GENENAME"),
              keytype="ENSEMBL")
# Have a look at the annotation
anno

```

However, we have a problem that the resulting data frame has more rows than our results table. This is due to the *one-to-many* relationships that often occur when mapping between various identifiers.

```{r}
dim(anno)
dim(results.ordered)
```

Such duplicated entries can be identified using the `duplicated` function. 

```{r}
dup_ids <- anno$ENSEMBL[duplicated(anno$ENSEMBL)]
filter(anno, ENSEMBL %in% dup_ids) %>% 
  arrange(ENSEMBL)

```

Fortunately, there are not too many so hopefully we won't lose too much information if we discard the entries that are duplicated. The first occurence of the duplicated ID will still be included in the table.

```{r}
anno <- select(org.Mm.eg.db,keys=results.ordered$ENSEMBL,
              columns=c("ENSEMBL","SYMBOL","GENENAME","ENTREZID"),
              keytype="ENSEMBL") %>% 
  filter(!duplicated(ENSEMBL))
dim(anno)
```


We can bind in the annotation information to the `results` data frame. 

```{r}
results.annotated <- left_join(results.ordered, anno,by="ENSEMBL")
results.annotated

```


We can save the results table using the `write.csv` function, which writes the results out to a csv file that you can open in excel.

```{r}
write.csv(results.annotated,file="virgin_vs_lactation_DESeq_annotated.csv",row.names=FALSE)
```


The gene names that we have discovered can now be added to any plots that we have created. However, we still have to use the Ensembl IDs when accessing the `DESeqDataSet` object.

```{r}
plotCounts(dds, "ENSMUSG00000000381",intgroup = c("Status"),main="Wap")
```


We have already seen the use of a heatmap as a quality assessment tool to visualise the relationship between samples in an experiment. Another common use-case for such a plot is to visualise the results of a differential expression analysis....

Here we will take the top 10 genes from the differential expression analysis and produce a heatmap. The default colour palette goes from low expression in blue to high expression in red, which is a good alternative to the traditional red/green heatmaps which are not suitable for those with forms of colour-blindness.

```{r}
library(pheatmap)
top_genes <- results.annotated$ENSEMBL[1:10]

vsd <- vst(dds)
pheatmap(assay(vsd)[top_genes,])


```

The heatmap is more informative if we add colours underneath the sample dendrogram to indicate which sample group each sample belongs to. This we can do by creating a data frame containing metadata for each of the samples in our dataset. With the `DESeq2` workflow we have already created such a data frame. We have to make sure the the rownames of the data frame are the same as the column names of the counts matrix.

```{r}
sampleInfo <- as.data.frame(colData(dds)[,c("Status","CellType")])

pheatmap(assay(vsd)[top_genes,],
         annotation_col = sampleInfo)
```

Any plot we create in RStudio can be saved as a png or pdf file. We use the `png` or `pdf` function to create a file for the plot to be saved into and run the rest of the code as normal. The plot does not get displayed in RStudio, but printed to the specified file. 

```{r}

png("heatmap_top10_genes.png",width=800,height=800)
pheatmap(assay(vsd)[top_genes,],
         annotation_col = sampleInfo)
# dev.off()
```


> ## Challenge 1{.challenge}
> 1. Repeat the same heatmap as above, but for the top 100 most differentially-expressed genes
> 2. Change the plot so that gene names are displayed rather than Entrez IDs
> 3. Save the plot to a pdf file
> HINT: check the help for the `pheatmap` function to see how column and row labels can be changed

### Accessing the sample or clusters

The heatmap displays relationships between samples and genes in our study as a useful visualisation. In this example we can easily identify which samples are most similar based on their expression patterns. However, for larger dataset this may be more problematic. We can extract data the sample relationships about if we manually perform the clustering steps used by `pheatmap`. First is to cluster the samples with the default distance matrix and clustering algorithms.

```{r}
mat <- assay(vsd)[top_genes,]
## Calculate the distance matrix between samples
d_samples <- dist(t(mat))

plot(hclust(d_samples))
## Do clustering and split the samples into 2 groups
```

We can then "cut" the dendrogram to give a set number of clusters. The groupings could then be associated with the sample metadata.

```{r}
clusters <- cutree(hclust(d_samples),k = 2)
clusters
table(clusters, colData(dds)$Status)

```

A similar approach would also work for the genes. The key difference that we don't transpose before calculating the distance matrix. For the small number of genes in the heatmap such a distance matrix can be computed, but it may be computationally-expensive for a large number of genes.

```{r}
d_genes <- dist(mat)

plot(hclust(d_genes))
## Do clustering and split the samples into 2 groups
cutree(hclust(d_genes),k = 2)

```


### Annotation with the biomaRt resource

The Bioconductor package have the convenience of being able to make queries offline. However, they are only available for certain organisms. If your organism does not have an `org.XX.eg.db` package listed on the Bioconductor annotation page (http://bioconductor.org/packages/release/BiocViews.html#___AnnotationData), an alternative is to use biomaRt which provides an interface to the popular biomart annotation resource. 

The first step is to find the name of a database that you want to connect to

```{r}
library(biomaRt)
listMarts()
ensembl=useMart("ENSEMBL_MART_ENSEMBL")
# list the available datasets (species). Replace mouse with the name of your organism
listDatasets(ensembl) %>% filter(grepl("Mouse",description))

```

```{r}
ensembl = useDataset("mmusculus_gene_ensembl", mart=ensembl)
```

Queries to `biomaRt` are constructed in a similar way to the queries we performed with the `org.Mm.eg.db` package. Instead of `keys` we have `filters`, and instead of `columns` we have attributes. The list of acceptable values is much more comprehensive that for the `org.Mm.eg.db` package.

```{r}
listFilters(ensembl) %>% 
    filter(grepl("ensembl",name))
```


```{r}
listAttributes(ensembl) %>% 
    filter(grepl("gene",name))
```

An advantage over the `org..` packages is that positional information can be retrieved

```{r}
attributeNames <- c('ensembl_gene_id', 'entrezgene', 'external_gene_name')

getBM(attributes = attributeNames,
      filters = "ensembl_gene_id",
      values=top_genes,
      mart=ensembl)
```

> ## Challenge 1{.challenge}
> 1. Use biomaRt to create an data frame containing the entrezgene, gene symbol and genomic coordinates (chromosome, start, end) for the Ensembl IDs in the DESeq2 results
> 2. Remove duplicates entries from the new data frame
> 3. Join the biomaRt annotation to the DESeq2 results to produce a data frame with differential expression results and annotation
> 4. Write the joined data frame to a csv file

```{r}

```



### Brief Introduction to ggplot2

The [`ggplot2`](http://ggplot2.tidyverse.org/) package has emerged as an attractive alternative to the traditional plots provided by base R. A full overview of all capabilities of the package is available from the [cheatsheet](https://www.rstudio.com/wp-content/uploads/2015/03/ggplot2-cheatsheet.pdf).

The `plotCounts` function that we saw earlier can be adjusted to return the data associated with the plot, which in-turn can be passed to a `ggplot2` function for customisable plots.

- `plot_data` is our data frame containing the variables we wish to plot
- `aes` creates a mpping between the variables in our data frame to the *aes*thetic proprties of the plot
    + the x-axis is mapped to `Status`, y-axis is mapped to `count`
- `geom_point` specifies the particular type of plot we want (in this case a scatter plot)
    + see [the cheatsheet](https://www.rstudio.com/wp-content/uploads/2015/03/ggplot2-cheatsheet.pdf) for other plot types


```{r}
plot_data <- plotCounts(dds, "ENSMUSG00000000381",intgroup = c("CellType","Status"),returnData = TRUE)
plot_data

ggplot(plot_data, aes(x = Status,y=log2(count),col=CellType)) + geom_point()
```

```{r}
ggplot(plot_data, aes(x = Status,y=log2(count),col=Status)) + geom_point()  + facet_wrap(~CellType)
```



A simple scatter plot, equivalent to `plotMA` from before, can be generated as follows:-

```{r,fig.width=12,fig.height=5}
library(ggplot2)
df <- results.annotated
ggplot(df, aes(x = log2(baseMean), y=log2FoldChange)) + geom_point() 

```

In brief:-


The real advantage of `ggplot2` is the ability to change the appearance of our plot by mapping other variables to aspects of the plot. For example, we could colour the points based on a p-value cut-off. The colours are automatically chosen by `ggplot2`, but we can specifiy particular values.

```{r,fig.width=12,fig.height=5}

ggplot(df, aes(x = log2(baseMean), y=log2FoldChange,col=padj<0.05)) + geom_point() 

ggplot(df, aes(x = log2(baseMean), y=log2FoldChange)) + geom_point(aes(color=padj<0.05),alpha=0.4) +  scale_colour_manual(values=c("black","red"))

```

Text can also be added to the plot to indicate the names of particular genes of interest. A column can be added to the data frame used for plotting with a text label that can be associated with each position on the plot. The plot would quickly become too cluttered if we added too many labels, but we can put empty strings `""` for genes that we don't want to plot.


```{r fig.wdith=12}
N <- 10

top_genes <- df$ENSEMBL[1:N]

df<- mutate(df, Label=ifelse(df$ENSEMBL %in% top_genes, SYMBOL, ""))

ggplot(df, aes(x = log2(baseMean), y=log2FoldChange)) + geom_point(aes(color=padj<0.05),alpha=0.4,size=0.5) +  scale_colour_manual(values=c("black","red")) +  geom_text(aes(label=Label),col="blue",alpha=0.8)


```



## Retrieving Genomic Locations


It might seem natural to add genomic locations to our annotation table, and possibly a bit odd that the `org.Mm.eg.db` package does not supply such mappings. In fact, there is a whole suite of package for performing this, and more-advanced queries that relate to the location of genes. These are listed on the Bioconductor [annotation page](http://bioconductor.org/packages/release/BiocViews.html#___AnnotationData) and have the prefix `TxDb.`

The package we will be using is `TxDb.Mmusculus.UCSC.mm10.knownGene`. Packages are available for other organisms and genome builds. It is even possible to *build your own database* if one does not exist. See `vignette("GenomicFeatures")` for details

```{r eval=FALSE}
source("http://www.bioconductor.org/biocLite.R")
biocLite("TxDb.Mmusculus.UCSC.mm10.knownGene")

## For Humans
biocLite("TxDb.Hsapiens.UCSC.hg19.knownGene")

```

We load the library in the usual fashion and create a new object to save some typing. As with the `org.` packages, we can query what columns are available with `columns`,

```{r message=FALSE}
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
tx <- TxDb.Mmusculus.UCSC.mm10.knownGene
columns(tx)
```

The `select` function is used in the same manner as the `org.Mm.eg.db` packages. 


> ## Challenge 2{.challenge}
>
> Use the TxDb.Mmusculus.UCSC.mm10.knownGene package to retrieve the exon coordinates for the genes `50916`, `110308`, `12293` 
>

```{r echo=FALSE,warning=FALSE,message=FALSE}


```

### Overview of GenomicRanges

One of the real strengths of the `txdb..` packages is the ability of interface with `GenomicRanges`, which is the object type used throughout Bioconductor [to manipulate Genomic Intervals](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3738458/pdf/pcbi.1003118.pdf). 

These object types permit us to perform common operations on intervals such as overlapping and counting. We can define the chromosome, start and end position of each region (also strand too, but not shown here).

```{r}
library(GenomicRanges)
simple.range <-GRanges("1", IRanges(start=1000,end=2000))
simple.range

```

We don't have to have all our ranges located on the same chromosome
```{r}
chrs <- c("chr13", "chr15","chr5")
start <- c(73000000, 101000000, 15000000)
end <- c(74000000,102000000, 16000000)

my.ranges <- GRanges(rep(chrs,3), 
                     IRanges(start=rep(start,each=3),
                             end = rep(end,each=3))
)

```

There are a number of useful functions for calculating properties of the data (such as *coverage* or sorting). Not so much for RNA-seq analysis, but `GenomicRanges` are used throughout Bioconductor for the analysis of NGS data. 

For instance, we can quickly identify overlapping regions between two `GenomicRanges`. However, we have to pay attention to the naming convention used for each object. `seqlevelsStyle` can 

```{r}
keys <- c("50916","110308","12293")
genePos <- select(tx, keys=keys,
       keytype = "GENEID",
       columns=c("EXONCHROM","EXONSTART","EXONEND")
      )
geneRanges <- GRanges(genePos$EXONCHROM, IRanges(genePos$EXONSTART,genePos$EXONEND), GENEID=genePos$GENEID)
geneRanges

findOverlaps(my.ranges,geneRanges)
seqlevelsStyle(geneRanges)
seqlevelsStyle(simple.range)

```


## Retrieving Gene Coordinates as GenomicRanges

As we saw above, it is quite straightforward to translate the output of a `select` query into a `GenomicFeatures` object. However, several convenience functions exist to retrieve the structure of every gene for a given organism in one object. 

The output of `exonsBy` is a list, where each item in the list is the exon co-ordinates of a particular gene. 

```{r}
exo <- exonsBy(tx,"gene")
exo
```

To access the structure of a particular gene, we can use the `[[` syntax with the name of the gene (Entrez gene ID) within quote marks. If we wanted to whole region that the gene spans we could use the `range` function.

```{r}
exo[["110308"]]
range(exo[["110308"]])
```


## Exporting tracks

It is also possible to save the results of a Bioconductor analysis in a browser to enable interactive analysis and integration with other data types, or sharing with collaborators. For instance, we might want a browser track to indicate where our differentially-expressed genes are located. We shall use the `bed` format to display these locations. We will annotate the ranges with information from our analysis such as the fold-change and significance.

First we create a data frame for just the DE genes.
```{r}
sigGenes <- results.annotated[which(results.annotated$padj < 0.05),]
sigGenes
```

At the moment, we have a GenomicFeatures object that represents every exon. However, we do not
need this level of granularity for the bed output, so we will collapse to a single region for each gene. First we the `range` function to obtain a single range for every gene and tranform to a more convenient object with `unlist`.
```{r}
exoRanges <- unlist(range(exo))
sigRegions <- exoRanges[na.omit(match(sigGenes$ENTREZID, names(exoRanges)))]
sigRegions
```

Rather than just representing the genomic locations, the .bed format is also able to colour each range
according to some property of the analysis (e.g. direction and magnitude of change) to help highlight
particular regions of interest. A score can also be displayed when a particular region is clicked-on.
A useful propery of GenomicRanges is that we can attach *metadata* to each range using the `mcols`
function. The metadata can be supplied in the form of a data frame.

```{r}
mcols(sigRegions) <- sigGenes[match(names(sigRegions), rownames(sigGenes)),]
sigRegions
```

The metadata we have added can also by used as a means to interrogate the ranges; as if the data were contained in a data frame.

```{r}
sigRegions[order(sigRegions$log2FoldChange,decreasing = TRUE)]
```

For visualisation purposes, we are going to restrict the data to genes that are located on chromosomes 1 to 19 and the sex chromosomes. This can be done with the `keepSeqLevels` function.

```{r}
seqlevels(sigRegions)
sigRegions <- keepSeqlevels(sigRegions, paste0("chr", c(1:19,"X","Y")),pruning.mode = "coarse")
```

We will now create a score from the p-values that will displayed under each region, and colour scheme
for the regions based on the fold-change. For the score we can use the $-log_{10}$ of the adjusted p-value as before



```{r}
Score <- -log10(sigRegions$padj)
```

`colorRampPalette` is a useful function in base R for constructing a palette between two extremes. **When choosing colour palettes, make sure they are colour blind friendly**. The red / green colour scheme traditionally-applied to microarrays is a ***bad*** choice.

We will also truncate the fold-changes to between -5 and 5 to and divide this range into 10 equal bins

```{r}
rbPal <-colorRampPalette(c("red", "blue"))
logfc <- pmax(sigRegions$log2FoldChange, -5)
logfc <- pmin(logfc , 5)

Col <- rbPal(10)[as.numeric(cut(logfc, breaks = 10))]
```

The colours and score have to be saved in the GRanges object as `score` and `itemRgb` columns respectively, and will be used to construct the browser track. The rtracklayer package can be used to import and export browsers tracks.

Now we can export the signifcant results from the DE analysis as a `.bed` track using `rtracklayer`. You can load the resulting file in IGV, if you wish.
```{r}
mcols(sigRegions)$score <- Score
mcols(sigRegions)$itemRgb <- Col
sigRegions
library(rtracklayer)
export(sigRegions , con = "topHits.bed")
```
