---
title: "RNA-seq analysis in R"
subtitle: "Differential Expression of RNA-seq data"
author: "Mark Dunning"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output:
  html_notebook:
    toc: yes
    toc_float: yes
  html_document:
    toc: yes
    toc_float: yes
minutes: 300
layout: page
bibliography: ref.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

**Original Authors: Belinda Phipson, Anna Trigos, Matt Ritchie, Maria Doyle, Harriet Dashnow, Charity Law**, **Stephane Ballereau, Oscar Rueda, Ashley Sawle**
Based on the course [RNAseq analysis in R](http://combine-australia.github.io/2016-05-11-RNAseq/) delivered on May 11/12th 2016 and modified by Cancer Research Uk Cambridge Centre for the [Functional Genomics Autumn School 2017](https://bioinformatics-core-shared-training.github.io/cruk-autumn-school-2017/)

## Resources and data files

This material has been created using the following resources:  
http://www.statsci.org/smyth/pubs/QLedgeRPreprint.pdf [@Lun2016]  
http://monashbioinformaticsplatform.github.io/RNAseq-DE-analysis-with-R/99-RNAseq_DE_analysis_with_R.html  
http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html

Data files downloaded from:  
ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE60nnn/GSE60450/suppl/GSE60450_Lactation-GenewiseCounts.txt.gz
http://bioinf.wehi.edu.au/software/MSigDB/mouse_c2_v5.rdata
http://bioinf.wehi.edu.au/software/MSigDB/mouse_H_v5.rdata

Data files:  
sampleinfo.txt  
GSE60450_Lactation-GenewiseCounts.txt  
mouse_c2_v5.rdata  
mouse_H_v5.rdata

Data files available from: [https://figshare.com/s/1d788fd384d33e913a2a](https://figshare.com/s/1d788fd384d33e913a2a)
You should download these files and place them in your `/data` directory.

## Differential expression with `DESeq2`

Now that we are happy that we have normalised the data and that the quality looks good, we can continue to testing for differentially expressed genes. There are a number of packages to analyse RNA-Seq data. Most people use `DESeq2` or `edgeR`. We will use `DESeq2` for the rest of this practical.

**First make sure we have all the objects and libraries loaded*

```{r}
library(DESeq2)
library(edgeR)

```


### Recap of pre-processing

The previous section walked-through the pre-processing and transformation of the count data. Here, for completeness, we list the minimal steps required to process the data prior to differential expression analysis.

```{r eval=FALSE}
library(DESeq2)
library(edgeR)
## Read the counts from the downloaded data
seqdata <- read.delim("data/GSE60450_Lactation-GenewiseCounts.txt", stringsAsFactors = FALSE)
#
# Remove first two columns from seqdata

countdata <- seqdata[,-(1:2)]

# Store EntrezGeneID as rownames
rownames(countdata) <- seqdata[,1]
countdata
colnames(countdata) <- substr(colnames(countdata), 1, 7)
countdata
## Calculate the Counts Per Million measure
myCPM <- cpm(countdata)
## Identify genes with at least 0.5 cpm in at least 2 samples
thresh <- myCPM > 0.5
keep <- rowSums(thresh) >= 2
# Subset the rows of countdata to keep the more highly expressed genes
counts.keep <- countdata[keep,]
## Convert to an edgeR object
sampleinfo <- read.delim("data/SampleInfo_Corrected.txt")

dds <- DESeqDataSetFromMatrix(counts.keep,
                              colData = sampleinfo,
                              design = ~CellType)

```

We also have the output of the pre-processing section saved as an R object if you didn't manage to complete these steps.

```{r}
## Only run if you didn't complete the previous section on pre-processing
load("Robjects/preprocessing.Rdata")
```


### Overview of testing

As we saw in the R introduction, there are many options for performing a test to assess the significance of data from different sample groups. Taking a toy example:-

```{r}
gene_counts <- c(rnorm(10,mean = 2,sd=1),
                 rnorm(10, mean=4,sd=1))

group <- c(rep("Treated",10),
           rep("Control",10))

boxplot(gene_counts ~ group)

```

However, we cannot naively apply such a test to RNA-seq data that are *count-based*. Much work in the microarray era was carried out on statistical analysis of gene expression data with the most-popular approach being that of the limma Bioconductor package. This package has been cited thousands of time and applies a *linear model* approach to the analysis of gene expression data. The [user guide](https://www.bioconductor.org/packages/devel/bioc/vignettes/limma/inst/doc/usersguide.pdf) is extensive and has many examples of applying linear model design to expression data. It is even possible to use limma on RNA-seq data provided that the data are [appropriately transformed](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29). 

For this workflow we are going to use `DESeq2` which uses a negative binomial distribution to model the data.


### Differential Expression with DESeq2

In `DESeq2`, a differential expression analysis can be performed by using the `DESeq` function. This function requires a single parameter which is the `DESeqDataSet` object generated in the pre-processing steps describes. We have previously defined the test condition using the `design` argument when we created the object. This can be checked using the `design` function.

The function runs a couple of processing steps automatically to adjust for different library size and gene-wise variabiliy, which you can read about in the [DESeq2 vignette)[http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#the-deseq2-model].

```{r}
colData(dds)
design(dds) 
de <- DESeq(dds)

```

Typically we decide the design for the analysis when we create the DESeq2 objects, but it can be modified prior to the differential expression analysis

```{r}
design(dds) <- ~CellType
```


The results of the analysis can be obtained using the `results` function and displayed to the screen. Each row is a particular gene measured in the study (i.e. all genes in the organism being studies) and each column reports some aspect of the differential expression analysis for that gene. Note that all genes are reported. Not all genes will be statistically significant. At this stage the gene identifiers are not very informative, something we will fix in the next section. We also cannot easily identify genes that are differentially expressed.

```{r}
de <- DESeq(dds)
results.cellType <- as.data.frame(results(de))
results.cellType

```

> ## Challenge 1 {.challenge}
>
> 1. How many genes have an adjusted p-value less than 0.05?
> 2. Produce a data frame where the rows are ordered according to adjusted p-value
> 3. Write this new data frame to a `.csv` file
> 4. Re-run the analysis to find differentially-expressed genes between *Virgin* and *Lactate*
> HINT: you may need to use the `na.omit` function to ignore `NA` values in a vector
> HINT: you can use the `order` function that we saw in the R introduction to re-order rows...



```{r}


```

We can also incorporate the `log2FoldChange` values when we perform filtering 

```{r}

### Code to filter by log2 fold change
```


In this initial analyis `DESeq2` has automatically decided which member of our sample groups to use as our baseline( `basal` in this case) so that the log2 fold changes are reported with a positve value meaning higher expression in `luminal`. If we want to change this behaviour we can change the `contrast` argument in the `results` function


```{r}
## This should give the same as the table above
results(de, contrast=c("CellType","luminal","basal"))
## Changing the direction of the contrast
results(de, contrast=c("CellType","basal","luminal"))

```

If we change to performing differential expression analysis on the `Status` variable then there are various contrasts that can be made; `pregnant` vs `lactate`, `lactate` vs `virgin` etc. When the `results` function is run the table that is displayed is for the contrast `virgin vs lactate`. The `resultsNames` function can tell us which other contrasts we can access.


```{r}
dds.status <- dds
design(dds.status) <- ~Status
de.status <- DESeq(dds.status)
resultsNames(de.status)
results.status <- results(de.status)
```

A venn diagram is a common way of visualising the overlap between two genelists. We need to create a data frame where each column indicates whether each gene is differentially expressed in a particular contrast or not. To create such columns we can do a logical test on the adjusted p-values from our results tables.


```{r}
venn_data <- data.frame(CellType = results.cellType$padj<0.05,
                        Status = results.status$padj < 0.05)
library(limma)
vennDiagram(venn_data)
```


> ## Challenge 2 {.challenge}
>
> 1. Use a venn diagram to visualise the overlap in the genes found to be differentially expressed in the `pregnant vs lactate` and `virgin vs lactate` contrasts. 
> 2. How many genes are in common?

```{r}

```





### Fitting alternative models to the data

`DESEq2` allows for more complicated models to be fit to the data. For guidance on how to fit more complicated models you can consult the [DESeq2 vignette](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html), the [limma user guide](https://www.bioconductor.org/packages/devel/bioc/vignettes/limma/inst/doc/usersguide.pdf) or the Bioconductor mailing list. 

In particular, DESeq2 allows [multi-factor models](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#multi-factor-designs) which can account for other sources of variation in the data such as batches or gender.

Lets suppose that we wanted the different between virgin and lactatin individuals, but controlling for `CellType`. The design for such an analysis would be:-

```{r}
dds.mf <- dds
design(dds.mf) <- ~CellType+Status
de.mf <- DESeq(dds.mf)
results.mf <- results(de.mf,contrast=c("Status","virgin","lactate"))
results.mf
```



